README.md
# :construction: :crab: Hexbot :crab: :construction: <br> [![top language: rust]][rust-lang] ![rust 2018] ![rustc 1.39+] [![GPLv3+]][COPYING] ![tag]

[top language: rust]: https://img.shields.io/github/languages/top/rusty-snake/hexbot.svg?logo=rust
[rust-lang]: https://www.rust-lang.org/
[rustc 1.39+]: https://img.shields.io/badge/rustc-1.39+-blue.svg?logo=rust
[rust 2018]: https://img.shields.io/badge/rust--edition-2018-blue.svg?logo=rust
[GPLv3+]: https://img.shields.io/github/license/rusty-snake/hexbot.svg?color=darkred
[COPYING]: COPYING
[tag]: https://img.shields.io/github/tag/rusty-snake/hexbot.svg

[![Travis CI Status](https://badgen.net/travis/rusty-snake/hexbot/master?icon=travis&label=build)](https://travis-ci.com/rusty-snake/hexbot)
[![Dependabot Status](https://api.dependabot.com/badges/status?host=github&repo=rusty-snake/hexbot)](https://dependabot.com)
![actively developed](https://badgen.net/badge/maintenance/actively-developed/green)

My solution for: https://noopschallenge.com/challenges/hexbot

## Getting Started

### Install [rust](https://www.rust-lang.org/)

The canonical way to [install rust] is to use rustup:

```bash
curl https://sh.rustup.rs -sSf | sh
```

You need this on distros who do not have a supported rustc version in their packet sources.
e.g. Debian Stable, Mint, openSUSE Leap, Ubuntu, ...

On distros which have the current rustc version in their packet sources, it can be used.

| Distro                  | command                     |
| ----------------------- | --------------------------- |
| Arch Linux              | `sudo pacman -S rust`       |
| Debian Testing/Unstable | `sudo apt install cargo`    |
| Fedora                  | `sudo dnf install cargo`    |
| openSUSE Tumbleweed     | `sudo zypper install cargo` |

[install rust]: https://www.rust-lang.org/tools/install

### Get the source code

```bash
git clone https://github.com/rusty-snake/hexbot.git
cd hexbot
```

### Compile & Run

```
$ cargo run --release
    Updating crates.io index
   ...
   Compiling hexbot v0.0.11 (/home/rusty-snake/hexbot)
    Finished release [optimized] target(s) in 4m 2s
     Running `target/release/hexbot`
===== Hexbot =====
Should the count parameter be added? [yes|no] yes
What value should count have? [1-1000] 5
Should the width and height parameters be added? [yes|no] yes
What value should width have? [10-100,000] 500
What value should height have? [10-100,000] 500
[#E46AF7-(370|226), #FAFD72-(334|70), #6C1882-(440|490), #ECC44A-(451|181), #F528DF-(151|72)]
```

#### compile only

```
$ cargo build --release
```

#### documentation

```
$ cargo doc --no-deps --open
```

## Hacking

Using the hexbot library in your own project.

`Cargo.toml`:
```toml
[dependencies]
hexbot = { git = "https://github.com/rusty-snake/hexbot", tag = "v0.0.12" }
futures = "0.3"
```

`src/main.rs`:
```rust
use hexbot::{Count, Hexbot, Seed, WidthHeight};
use futures::executor::block_on;

fn main() {
    let hb = block_on(Hexbot::fetch(
        Count::no(),
        WidthHeight::no(),
        &Seed::no()
    )).expect("Fetching failed");
    println!("Hello from Hexbot: {}", hb);
}
```

For the next steps, see the [documentation](#documentation).

## Changelog

```markdown
## [0.0.12] - 2020-02-01
### Removed
 * `WithCoordinates` and `CoordinatesLimit`

### Changed
 * make `Hexbot::fetch` async
 * min rustc: 1.39.0

[0.0.12]: https://github.com/rusty-snake/hexbot/tree/v0.0.12
```

For the full Changelog see [CHANGELOG.md](CHANGELOG.md).

## License

[GPL-3.0-or-later](COPYING)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.12] - 2020-02-01
### Removed
 * `WithCoordinates` and `CoordinatesLimit`

### Changed
 * make `Hexbot::fetch` async
 * min rustc: 1.39.0

## [0.0.11] - 2019-10-13
### Added
 * `Seed::add`
 * impl ops::{Add,Sub} for Coordinates
 * `hexbot::WidthHeight` improved replacement for `WithCoordinates` and `CoordinatesLimit`

### Changed
 * `Hexbot::fetch` accept both `WithCoordinates` and `WidthHeight`
 * old:`SeedError::NoColor` new:`SeedError::NoColor(i32)`

### Deprecated
 * `WithCoordinates` use `WidthHeight`
 * `CoordinatesLimit` use `WidthHeight`

## [0.0.10] - 2019-09-28
### Changed
 * min rustc: 1.37.0

### Added
  * library to use the hexbot API

### Removed
  * feature: `ErrorDescription`
  * `src/request_api.rs`

## [0.0.9] - 2019-09-04
maintenance release
 * fixing various non-code things
 * updating dependencies

## [0.0.8]
### Added
 * Support for error and message response
 * `.rustfmt.toml`

### Changed
 * Switch from openssl to rustls

## [0.0.7]
### Added
 * more tests.
 * support for hexbots `seed` parameter
   * add `seed: Option<&[i32]>` to `fetch()` and `fetch_with_coordinates()`
 * more Errors
   * `Fmt(fmt::Error)`
   * `EmptySeed`
   * `SeedToLong`
   * `InvalidSeedColor`

## [0.0.6]
### Added
 * `Hexbot.clone()` (derive trait).
 * support for the width and height parameters of hexbot.
   * `Hexbot.coordinates()`
   * `Hexbot.has_coordinates()`
   * `request_api::fetch_with_coordinates()`
   * `Error::WidthHeightOutOfRange`

### Changed
 * `fmt::Display` for `Hexbot`.
 * Improved docs & Hacking.

## [0.0.5]
### Added
 * custom error type.
 * more tests.
 * `request_api::Hexbot.colors()`.
 * support for the count parameter of hexbot.

### Changed
 * output of `fmt::Display` for Hexbot.
 * `request_api::fetch()`
   * old: `request_api::fetch() -> Result<Hexbot, reqwest::Error>`
   * new: `request_api::fetch(count: i32) -> Result<Hexbot, Error>`

### Removed
 * `request_api::Hexbot.color()`.

## [0.0.4]
### Added
 * documentation & hacking.

### Changed
 * Function names:
   * `get_color()` -> `color()`
   * `get_hexbot()` -> `fetch()`
 * Use `tint::Color` instead of `String` for colors.

## [0.0.3]
### Added
 * tests (see c461bb9).

## [0.0.2]
### Added
 - `get_color()` methode to `request_api::Hexbot`.
 - `impl`lementation for `fmt::Display` to `request_api::Hexbot`.

### Removed
 - `pub`lic fields from `request_api::Hexbot`.

## [0.0.1]
### Added
 - Support for requesting, parsing and printing a hexbot request without parameters.
 - All stuff around a project like README, LICENSE, .gitignore, ...

[0.0.12]: https://github.com/rusty-snake/hexbot/tree/v0.0.12
[0.0.11]: https://github.com/rusty-snake/hexbot/tree/v0.0.11
[0.0.10]: https://github.com/rusty-snake/hexbot/tree/v0.0.10
[0.0.9]: https://github.com/rusty-snake/hexbot/tree/v0.0.9
[0.0.8]: https://github.com/rusty-snake/hexbot/tree/v0.0.8
[0.0.7]: https://github.com/rusty-snake/hexbot/tree/v0.0.7
[0.0.6]: https://github.com/rusty-snake/hexbot/tree/v0.0.6
[0.0.5]: https://github.com/rusty-snake/hexbot/tree/v0.0.5
[0.0.4]: https://github.com/rusty-snake/hexbot/tree/v0.0.4
[0.0.3]: https://github.com/rusty-snake/hexbot/tree/v0.0.3
[0.0.2]: https://github.com/rusty-snake/hexbot/tree/v0.0.2
[0.0.1]: https://github.com/rusty-snake/hexbot/tree/v0.0.1


src/main.rs
/*
 * Copyright © 2019,2020 rusty-snake <print_hello_world+License@protonmail.com>
 *
 * This file is part of rusty-snake's hexbot solution
 *
 * rusty-snake's hexbot solution is free software: you can redistribute
 * it and/or modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * rusty-snake's hexbot solution is distributed in the hope that it
 * will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use hexbot::{Count, Hexbot, Seed, WidthHeight};
use std::io;
use std::io::Write;

fn input(prompt: &str) -> io::Result<String> {
    let mut buffer = String::new();
    let mut stdout = io::stdout();
    let stdin = io::stdin();
    stdout.write_all(prompt.as_bytes())?;
    stdout.flush()?;
    stdin.read_line(&mut buffer)?;
    Ok(buffer.trim().to_string())
}

fn ask_bool(question: &str) -> io::Result<bool> {
    let answer = input(question)?.to_lowercase();
    if answer == "y" || answer == "yes" {
        Ok(true)
    } else if answer == "n" || answer == "no" {
        Ok(false)
    } else {
        println!("Sorry, try again.");
        ask_bool(question)
    }
}

fn ask_i32(question: &str) -> io::Result<i32> {
    if let Ok(answer) = input(question)?.parse() {
        Ok(answer)
    } else {
        println!("Sorry, try again.");
        ask_i32(question)
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("===== Hexbot =====");
    let count = if ask_bool("Should the count parameter be added? [yes|no] ")? {
        loop {
            match Count::yes(ask_i32("What value should count have? [1-1000] ")?) {
                Ok(count) => break count,
                Err(_) => println!("This is a invalid count, try again."),
            }
        }
    } else {
        Count::no()
    };
    let with_coordinates =
        if ask_bool("Should the width and height parameters be added? [yes|no] ")? {
            loop {
                match WidthHeight::yes(
                    ask_i32("What value should width have? [10-100,000] ")?,
                    ask_i32("What value should height have? [10-100,000] ")?,
                ) {
                    Ok(limit) => break limit,
                    Err(_) => println!("This is a invalid width/height, try again."),
                }
            }
        } else {
            WidthHeight::no()
        };
    let hb = Hexbot::fetch(count, with_coordinates, &Seed::no()).await.expect("Fetching failed");
    println!("{}", hb);
    Ok(())
}
